File: menu_contextual/lib/controller/user_controller.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/user_model.dart';

class UserService {
  final String baseUrl = "http://localhost:8080/api/users";

  // Registrar un nuevo usuario
  Future<UserModel> registerUser(UserModel user) async {
    final response = await http.post(
      Uri.parse('$baseUrl/register'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(user.toMap()),
    );
    if (response.statusCode == 201) {
      return UserModel.fromMap(json.decode(response.body));
    } else {
      throw Exception('Error al registrar usuario');
    }
  }

  // Login de usuario
  Future<UserModel> loginUser(String correo, String password) async {
    final response = await http.post(
      Uri.parse('$baseUrl/login'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({'correo': correo, 'password': password}),
    );
    if (response.statusCode == 200) {
      return UserModel.fromMap(json.decode(response.body));
    } else {
      throw Exception('Error al iniciar sesión');
    }
  }

  // Obtener un usuario por ID
  Future<UserModel> getUserById(String userId) async {
    final response = await http.get(Uri.parse('$baseUrl/$userId'));
    if (response.statusCode == 200) {
      return UserModel.fromMap(json.decode(response.body));
    } else {
      throw Exception('Usuario no encontrado');
    }
  }

  // Actualizar un usuario
  Future<UserModel> updateUser(String userId, UserModel updatedUser) async {
    final response = await http.put(
      Uri.parse('$baseUrl/$userId'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(updatedUser.toMap()),
    );
    if (response.statusCode == 200) {
      return UserModel.fromMap(json.decode(response.body));
    } else {
      throw Exception('Error al actualizar usuario');
    }
  }

  // Eliminar un usuario
  Future<void> deleteUser(String userId) async {
    final response = await http.delete(Uri.parse('$baseUrl/$userId'));
    if (response.statusCode != 200) {
      throw Exception('Error al eliminar usuario');
    }
  }
}

File: backend/backend-springboot/src/main/java/ec/edu/espe/bancaMovil/controller/UserController.java
package ec.edu.espe.bancaMovil.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import ec.edu.espe.bancaMovil.model.User;
import ec.edu.espe.bancaMovil.service.UserService;

@RestController
@RequestMapping("/api/users")
public class UserController {
    @Autowired
    private UserService userService;

    // ✅ REGISTRAR USUARIO - Devuelve 201 (CREATED)
    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody User user) {
        try {
            User newUser = userService.registerUser(user);
            return ResponseEntity.status(HttpStatus.CREATED).body(newUser);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Error al registrar usuario: " + e.getMessage());
        }
    }

    // ✅ INICIAR SESIÓN - Devuelve 200 (OK) o 401 (UNAUTHORIZED)
    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody User user) {
        try {
            User loggedUser = userService.loginUser(user.getCorreo(), user.getPassword());
            if (loggedUser != null) {
                return ResponseEntity.ok(loggedUser);
            } else {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Correo o contraseña incorrectos");
            }
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Error al iniciar sesión: " + e.getMessage());
        }
    }

    // ✅ OBTENER USUARIO POR ID - Devuelve 200 (OK) o 404 (NOT FOUND)
    @GetMapping("/{userId}")
    public ResponseEntity<?> getUserById(@PathVariable Long userId) {
        try {
            User user = userService.getUserById(userId);
            return ResponseEntity.ok(user);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body("Usuario no encontrado: " + e.getMessage());
        }
    }

    // ✅ ACTUALIZAR USUARIO - Devuelve 200 (OK) o 404 (NOT FOUND)
    @PutMapping("/{userId}")
    public ResponseEntity<?> updateUser(@PathVariable Long userId, @RequestBody User updatedUser) {
        try {
            User user = userService.updateUser(userId, updatedUser);
            return ResponseEntity.ok(user);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body("Error al actualizar usuario: " + e.getMessage());
        }
    }

    // ✅ ELIMINAR USUARIO - Devuelve 204 (NO CONTENT) o 404 (NOT FOUND)
    @DeleteMapping("/{userId}")
    public ResponseEntity<?> deleteUser(@PathVariable Long userId) {
        try {
            userService.deleteUser(userId);
            return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body("Error al eliminar usuario: " + e.getMessage());
        }
    }
}
